File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/adapter/PostsFilter.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.adamantcheese.chan.ui.adapter;

import android.text.TextUtils;

import com.github.adamantcheese.chan.core.database.DatabaseManager;
import com.github.adamantcheese.chan.core.manager.SiteManager;
import com.github.adamantcheese.chan.core.model.Post;
import com.github.adamantcheese.chan.core.model.PostImage;
import com.github.adamantcheese.chan.core.model.PostIndexed;
import com.github.adamantcheese.chan.core.site.Site;
import com.github.adamantcheese.model.data.descriptor.ChanDescriptor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

import javax.inject.Inject;

import static com.github.adamantcheese.chan.Chan.inject;

public class PostsFilter {
    private static final Comparator<Post> IMAGE_COMPARATOR = (lhs, rhs) -> rhs.getThreadImagesCount() - lhs.getThreadImagesCount();
    private static final Comparator<Post> REPLY_COMPARATOR = (lhs, rhs) -> rhs.getTotalRepliesCount() - lhs.getTotalRepliesCount();
    private static final Comparator<Post> NEWEST_COMPARATOR = (lhs, rhs) -> (int) (rhs.time - lhs.time);
    private static final Comparator<Post> OLDEST_COMPARATOR = (lhs, rhs) -> (int) (lhs.time - rhs.time);

    private static final Comparator<Post> MODIFIED_COMPARATOR =
            (lhs, rhs) -> (int) (rhs.getLastModified() - lhs.getLastModified());

    private static final Comparator<Post> THREAD_ACTIVITY_COMPARATOR = (lhs, rhs) -> {
        long currentTimeSeconds = System.currentTimeMillis() / 1000;

        //we can't divide by zero, but we can divide by the smallest thing that's closest to 0 instead
        long score1 = (long) ((currentTimeSeconds - lhs.time) / (lhs.getTotalRepliesCount() != 0
                ? lhs.getTotalRepliesCount()
                : Float.MIN_NORMAL));
        long score2 = (long) ((currentTimeSeconds - rhs.time) / (rhs.getTotalRepliesCount() != 0
                ? rhs.getTotalRepliesCount()
                : Float.MIN_NORMAL));

        return Long.compare(score1, score2);
    };

    @Inject
    DatabaseManager databaseManager;
    @Inject
    SiteManager siteManager;

    private Order order;
    private String query;

    public PostsFilter(Order order, String query) {
        this.order = order;
        this.query = query;
        inject(this);
    }

    public List<PostIndexed> apply(List<Post> original, ChanDescriptor chanDescriptor) {
        Site site = siteManager.bySiteDescriptor(chanDescriptor.siteDescriptor());
        if (site == null) {
            return Collections.emptyList();
        }

        String siteName = site.name();
        String board = chanDescriptor.boardCode();

        List<Post> posts = new ArrayList<>(original);
        Map<Long, Integer> originalPostIndexes = calculateOriginalPostIndexes(original);

        if (order != PostsFilter.Order.BUMP) {
            processOrder(posts);
        }

        if (!TextUtils.isEmpty(query)) {
            processSearch(posts);
        }

        // Process hidden by filter and post/thread hiding
        List<Post> retainedPosts = databaseManager.getDatabaseHideManager().filterHiddenPosts(
                posts,
                siteName,
                board
        );

        List<PostIndexed> indexedPosts = new ArrayList<>(retainedPosts.size());

        for (int currentPostIndex = 0; currentPostIndex < retainedPosts.size(); currentPostIndex++) {
            Post retainedPost = retainedPosts.get(currentPostIndex);
            int realIndex = Objects.requireNonNull(originalPostIndexes.get(retainedPost.no));
            indexedPosts.add(new PostIndexed(retainedPost, currentPostIndex, realIndex));
        }

        return indexedPosts;
    }

    private Map<Long, Integer> calculateOriginalPostIndexes(List<Post> original) {
        if (original.isEmpty()) {
            return Collections.emptyMap();
        }

        Map<Long, Integer> originalPostIndexes = new HashMap<>(original.size());

        for (int postIndex = 0; postIndex < original.size(); postIndex++) {
            Post post = original.get(postIndex);

            originalPostIndexes.put(post.no, postIndex);
        }

        return originalPostIndexes;
    }

    private void processOrder(List<Post> posts) {
        switch (order) {
            case IMAGE:
                Collections.sort(posts, IMAGE_COMPARATOR);
                break;
            case REPLY:
                Collections.sort(posts, REPLY_COMPARATOR);
                break;
            case NEWEST:
                Collections.sort(posts, NEWEST_COMPARATOR);
                break;
            case OLDEST:
                Collections.sort(posts, OLDEST_COMPARATOR);
                break;
            case MODIFIED:
                Collections.sort(posts, MODIFIED_COMPARATOR);
                break;
            case ACTIVITY:
                Collections.sort(posts, THREAD_ACTIVITY_COMPARATOR);
                break;
        }
    }

    private void processSearch(List<Post> posts) {
        String lowerQuery = query.toLowerCase(Locale.ENGLISH);

        boolean add;
        Iterator<Post> iterator = posts.iterator();

        while (iterator.hasNext()) {
            Post item = iterator.next();
            add = false;

            if (item.getComment().toString().toLowerCase(Locale.ENGLISH).contains(lowerQuery)) {
                add = true;
            } else if (item.subject != null
                    && item.subject.toString().toLowerCase(Locale.ENGLISH).contains(lowerQuery)) {
                add = true;
            } else if (item.name.toLowerCase(Locale.ENGLISH).contains(lowerQuery)) {
                add = true;
            } else if (item.getPostImagesCount() > 0) {
                for (PostImage image : item.getPostImages()) {
                    if (image.filename != null
                            && image.filename.toLowerCase(Locale.ENGLISH).contains(lowerQuery)) {
                        add = true;
                    }
                }
            }

            if (!add) {
                iterator.remove();
            }
        }
    }

    public enum Order {
        BUMP("bump"),
        REPLY("reply"),
        IMAGE("image"),
        NEWEST("newest"),
        OLDEST("oldest"),
        MODIFIED("modified"),
        ACTIVITY("activity");

        public String orderName;

        Order(String storeName) {
            this.orderName = storeName;
        }

        public static Order find(String name) {
            for (Order mode : Order.values()) {
                if (mode.orderName.equals(name)) {
                    return mode;
                }
            }
            return null;
        }

        public static boolean isNotBumpOrder(String orderString) {
            Order o = find(orderString);
            return !BUMP.equals(o);
        }
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/adapter/PostsFilter.kt
package com.github.k1rakishou.chan.ui.adapter

import com.github.k1rakishou.chan.core.helper.ChanLoadProgressEvent
import com.github.k1rakishou.chan.core.helper.ChanLoadProgressNotifier
import com.github.k1rakishou.chan.core.helper.PostHideHelper
import com.github.k1rakishou.common.mutableListWithCap
import com.github.k1rakishou.core_logger.Logger
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor
import com.github.k1rakishou.model.data.descriptor.PostDescriptor
import com.github.k1rakishou.model.data.post.ChanOriginalPost
import com.github.k1rakishou.model.data.post.ChanPost
import com.github.k1rakishou.model.data.post.PostIndexed
import com.github.k1rakishou.v2.KurobaSettings
import java.util.Collections

class PostsFilter(
  private val chanLoadProgressNotifier: ChanLoadProgressNotifier,
  private val postHideHelper: PostHideHelper,
  private val catalogSortingOrder: CatalogSortingOrder
) {

  suspend fun applyFilter(
    chanDescriptor: ChanDescriptor,
    posts: MutableList<ChanPost>,
    additionalPostsToReparse: MutableSet<PostDescriptor>
  ): List<PostIndexed> {
    chanLoadProgressNotifier.sendProgressEvent(
      ChanLoadProgressEvent.ApplyingFilters(
        chanDescriptor = chanDescriptor,
        postHidesCount = postHideHelper.countPostHides(posts),
        postFiltersCount = postHideHelper.countMatchedFilters(posts)
      )
    )

    if (catalogSortingOrder != CatalogSortingOrder.BUMP && chanDescriptor is ChanDescriptor.ICatalogDescriptor) {
      processOrder(catalogSortingOrder, posts as MutableList<ChanOriginalPost>)
    }

    // Process hidden by filter and post/thread hiding
    val retainedPosts = postHideHelper.processPostFilters(chanDescriptor, posts, additionalPostsToReparse)
      .safeUnwrap { error ->
        Logger.e(TAG, "postHideHelper.filterHiddenPosts error", error)
        return emptyList()
      }

    val indexedPosts = mutableListWithCap<PostIndexed>(retainedPosts.size)

    for ((index, retainedPost) in retainedPosts.withIndex()) {
      indexedPosts.add(PostIndexed(retainedPost, index))
    }

    return indexedPosts
  }

  enum class CatalogSortingOrder(val orderName: String) {
    BUMP("bump"),
    REPLY("reply"),
    IMAGE("image"),
    NEWEST("newest"),
    OLDEST("oldest"),
    MODIFIED("modified"),
    ACTIVITY("activity");

    val isBump: Boolean
      get() = this == BUMP

    companion object {
      fun current(kurobaSettings: KurobaSettings): CatalogSortingOrder {
        val borderOrderName = kurobaSettings.application.boardOrder.readBlocking()

        return entries.firstOrNull { it.orderName == borderOrderName }
          ?: BUMP
      }
    }
  }

  companion object {
    private const val TAG = "PostsFilter"

    fun processOrder(catalogSortingOrder: CatalogSortingOrder, posts: List<ChanOriginalPost>) {
      when (catalogSortingOrder) {
        CatalogSortingOrder.IMAGE -> Collections.sort(posts, IMAGE_COMPARATOR)
        CatalogSortingOrder.REPLY -> Collections.sort(posts, REPLY_COMPARATOR)
        CatalogSortingOrder.NEWEST -> Collections.sort(posts, NEWEST_COMPARATOR)
        CatalogSortingOrder.OLDEST -> Collections.sort(posts, OLDEST_COMPARATOR)
        CatalogSortingOrder.MODIFIED -> Collections.sort(posts, MODIFIED_COMPARATOR)
        CatalogSortingOrder.ACTIVITY -> Collections.sort(posts, THREAD_ACTIVITY_COMPARATOR)
        CatalogSortingOrder.BUMP -> {
          // no-op
        }
      }
    }

    private val IMAGE_COMPARATOR = Comparator<ChanOriginalPost> { lhs, rhs ->
      rhs.catalogImagesCount - lhs.catalogImagesCount
    }

    private val REPLY_COMPARATOR = Comparator<ChanOriginalPost> { lhs, rhs ->
      rhs.catalogRepliesCount - lhs.catalogRepliesCount
    }

    private val NEWEST_COMPARATOR = Comparator<ChanOriginalPost> { lhs, rhs ->
      (rhs.timestamp - lhs.timestamp).toInt()
    }

    private val OLDEST_COMPARATOR = Comparator<ChanOriginalPost> { lhs, rhs ->
      (lhs.timestamp - rhs.timestamp).toInt()
    }

    private val MODIFIED_COMPARATOR = Comparator<ChanOriginalPost> { lhs, rhs ->
      (rhs.lastModified - lhs.lastModified).toInt()
    }

    private val THREAD_ACTIVITY_COMPARATOR = Comparator<ChanOriginalPost> { lhs, rhs ->
      val currentTimeSeconds = System.currentTimeMillis() / 1000
      // we can't divide by zero, but we can divide by the smallest thing that's closest to 0 instead
      val eps = 0.0001f

      val lhsDivider = if (lhs.catalogRepliesCount > 0) {
        lhs.catalogRepliesCount.toFloat()
      } else {
        eps
      }

      val rhsDivider = if (rhs.catalogRepliesCount > 0) {
        rhs.catalogRepliesCount.toFloat()
      } else {
        eps
      }

      val score1 = ((currentTimeSeconds - lhs.timestamp).toFloat() / lhsDivider).toLong()
      val score2 = ((currentTimeSeconds - rhs.timestamp).toFloat() / rhsDivider).toLong()

      return@Comparator score1.compareTo(score2)
    }
  }

}